Reworked Base texture class

Quaker762 opened the pull request:
Textures can now be bound for operations by OpenGL, as well as the
ability to be loaded by the new Virtual File System we have. Textures
are stored as their pixel array in a vector.

Currently, 8-bit textures aren't supported because for reasons that
elude me, they are palletted, and I have no clue as to where the
pallete is (or what data it contains).

The header was also fixed up, as it was missing a 16-bit unused value,
which was distorting the number of bytes passed to the Virtual File
System to load the texture.

Quaker762 committed:
Reworked Base texture class

Textures can now be bound for operations by OpenGL, as well as the
ability to be loaded by the new Virtual File System we have. Textures
are stored as their pixel array in a vector.

Currently, 8-bit textures aren't supported because for reasons that
elude me, they are palletted, and I have no clue as to where the
pallete is (or what data it contains).

The header was also fixed up, as it was missing a 16-bit unused value,
which was distorting the number of bytes passed to the Virtual File
System to load the texture.

z33ky commented on source/SH3/graphics/texture.cpp:
```diff
@@ -4,44 +4,57 @@
+        Log(LogLevel::WARN, "sh3_texture::Load( ): Header only read %d bytes of data! It should have been %d", ret, sizeof(sh3_texture_header));
+
+    if(header.bpp == 8)
+        die("8-bit textures are currently unsupported!");
```
I think `die()` is a little drastic. We should warn here and get things rendering, even if it's messed up at the moment.
We could also check for other bpps we don't expect.

Quaker762 replied:
> I think die() is a little drastic

I thought so too, but the screen will end up looking like vomit [as you can see here](http://imgur.com/a/1lxeN). I'll remove it and log for now though.

z33ky commented on source/SH3/graphics/texture.cpp:
```diff
@@ -4,44 +4,57 @@
+    sh3_arc_vfile file(mft, filename);
+    sh3_arc_vfile::read_error e;
+
+    int ret = file.ReadData(&header, sizeof(sh3_texture_header), e); // Read in the header
```
Should be `std::size_t`.
I"m also in favor of using `sizeof(header)` (generally `sizeof(variable)` where applicable) because the intention here is to read the size of the variable `header`, which just happens to be a `sh3_texture_header` here.

Quaker762 replied:
> I"m also in favor of using sizeof(header) (generally sizeof(variable)

I'll do this from now on then. I think the whole `sizeof(typename)` has stuck from not only books I've read, but other peoples' code I've looked through.

z33ky replied:
This is still missing.

Quaker762 replied:
Fixed

z33ky commented on source/SH3/graphics/texture.cpp:
```diff
@@ -4,44 +4,57 @@
-    {
-        die("sh3_graphics::texture::Load( ): Unable to find file %s!", filename.c_str());
-    }
+    data.resize(header.texSize);
```
We could check that `texSize == texWidth * texHeight * bpp`. That should apply, right?
If it doesn't we should make sure that there's no out-of-bounds read from `data`.

Quaker762 replied:
As in an assertion??

z33ky replied:
Assertions should only be present in debug mode, so I think this should be a regular check & log.

z33ky replied:
This is still missing.

z33ky commented on source/SH3/graphics/texture.cpp:
```diff
@@ -4,44 +4,57 @@
+    // Now actually create a logical texture with OpenGL on the gpu
+    glGenTextures(1, &tex);
+
+    glBindTexture(GL_TEXTURE_2D, tex);
```
This could use `Bind()` and later `Unbind()`. Though apart from being shorter I don't see much benefit.

Quaker762 replied:
> This could use Bind() and later Unbind()

I have this weird habit of not using class functions inside other class functions... Strange I know..

z33ky reviewed (changes requested):
The texture also needs to be deleted at some point.

Have you considered bindless textures? I don't know if that is something we do want to use, it's just something that I've heard about and lingers in my mind.

Quaker762 commented:
> The texture also needs to be deleted at some point.

[Lmao](https://cdn.meme.am/cache/instances/folder610/500x/75266610.jpg)

> Have you considered bindless textures? I don't know if that is something we do want to use, it's just something that I've heard about and lingers in my mind.

I had never even heard of these until I googled them. It seems a little bit weird though (and the Khronos site even mentions it's a bit complex [here](https://www.khronos.org/opengl/wiki/Bindless_Texture))
I feel like binding makes a little bit more sense. You bind the texture you want to do something on, perform it, and then let it free. They'd be cool to look into and implement though, as long as they're not overly complex haha.

Quaker762 commented:
Also I just realised I should add texture units into bind, so we can bind to different texture units for different textures [as it explained here](https://www.opengl.org/discussion_boards/showthread.php/173263-What-is-a-texture-unit)

z33ky commented on include/SH3/graphics/texture.hpp:
```diff
@@ -69,47 +69,36 @@ namespace sh3_graphics
+         *  @param textureUnit The texture unit we want to bind this texture to
          */
-        load_error Load(const std::string& filename, sh3_arc& arc);
+         void Bind(GLenum textureUnit);
```
1 space too many

z33ky commented on source/SH3/graphics/texture.cpp:
```diff
@@ -4,44 +4,62 @@
+        Log(LogLevel::WARN, "sh3_texture::Load( ): Header only read %d bytes of data! It should have been %d", ret, sizeof(sh3_texture_header));
+
+    if(header.bpp == 8)
+        die("8-bit textures are currently unsupported!");
```
`die()` and `Log()`?

Quaker762 replied:
Crap I forgot to change this...

z33ky commented on source/SH3/graphics/texture.cpp:
```diff
@@ -4,44 +4,62 @@
+    int ret = file.ReadData(&header, sizeof(sh3_texture_header), e); // Read in the header
+
+    if(ret != sizeof(sh3_texture_header))
+        Log(LogLevel::WARN, "sh3_texture::Load( ): Header only read %d bytes of data! It should have been %d", ret, sizeof(sh3_texture_header));
```
We may want to fallback on a dummy texture when this happens, since it would be very cumbersome to ensure we don't read uninitialized struct fields.
For now this would be a good place to `die()`, though add a `//TODO` that we want to use a fallback texture in the future.

z33ky commented:
> I feel like binding makes a little bit more sense. You bind the texture you want to do something on, perform it, and then let it free. They'd be cool to look into and implement though, as long as they're not overly complex haha.

I find the stateful implementation weirder than the stateless one. The stateless version is OOP-like `do_stuff(thingToDoStuffOn, parameters...);` whereas the stateful has this verbose pattern of
```
make_thing_active(thingToDoStuffOn);
do_stuff(parameters...);
//sometimes: make_thing_inactive(thingToDoStuffOn);
```
Even worse if you made some function-calls between `make_thing_active` and `do_stuff`, because you have to hope that `thingToDoStuffOn` is still active, or you need to re-activate it just in case.
At least this is my understanding of the concept, without having read anything about the actual API.

Quaker762 commented:
> I find the stateful implementation weirder than the stateless one

Aren't programming paradigms and semantics fun haha.

> Even worse if you made some function-calls between make_thing_active and do_stuff, because you have to hope that thingToDoStuffOn is still active, or you need to re-activate it just in case.

Very true, though the chance of that happening are pretty slim with the amount we're reviewing each others code. Though I do see the problem in this. OpenGL wont shit itself if this happens, so your calls just silently fail or even worse corrupt something.

> At least this is my understanding of the concept, without having read anything about the actual API.

It seems like it. The only other issue I have is this is an ARB extension, which I assume most modern graphics cards support, but the chances of this being run on someone's grandma's PC is pretty high haha.

z33ky commented:
> The only other issue I have is this is an ARB extension, which I assume most modern graphics cards support, but the chances of this being run on someone's grandma's PC is pretty high haha.

I might have an idea how we can support both in a uniform API. I'll try that when this PR is ~merged~ rebased ;).

Quaker762 commented:
> I'll try that when this PR is merged rebased ;)

What do you need me to rebase?? Or is this the new workflow method you were talking about in #72 haha?

z33ky commented:
> Or is this the new workflow method you were talking about in #72 haha?

Yeah. The rebase will happen automatically if there's no conflicts.

Quaker762 commented:
> The rebase will happen automatically if there's no conflicts.

How the hell does that work!?!?!

z33ky commented:
See https://github.com/Palm-Studios/sh3redux/pull/76#pullrequestreview-19780384

z33ky commented on source/SH3/graphics/texture.cpp:
```diff
@@ -4,44 +4,62 @@
+
+    glBindTexture(GL_TEXTURE_2D, tex);
+    glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, header.texWidth, header.texHeight, 0, GL_BGRA, GL_UNSIGNED_BYTE, &data[0]);
+    glBindTexture(GL_TEXTURE_2D, 0); // Unbind this texture
```
Should this restore the previously bound texture (can be obtained via `glGetIntegerv(GL_TEXTURE_BINDING_2D, &prevTex);` I think)?

Quaker762 replied:
I don't quite understand why we'd want to bind the previous texture? This only changes a few parameters for this texture during a load. When the actual game is running, textures are bound and unbound when they're drawn (or rather, bound to different texture units and drawn) .

z33ky replied:
You mean they'll be (re-)bound per frame? That won't be an issue then.

Quaker762 replied:
Yeah they have to be to be drawn. Basically (how I'm thinking it'll work), is each mesh has a texture (*.tex). Everything is loaded before the scene is rendered. During the render loop, we transverse a list containing each mesh. Its VBO and texture is bound and drawn etc. 

When you generate a texture I'm pretty sure it's recommended you unbind it. Otherwise you could add a bunch of stuff to another texture without realising it.

Quaker762 committed:
Removed data vector from texture class

Once we upload the texture data to the Graphics Card, it is a waste of
conventional memory to keep this around...

z33ky commented on source/SH3/graphics/texture.cpp:
```diff
@@ -4,44 +4,69 @@
-    if(arc.LoadFile(filename, pixbuff) == ARC_FILE_NOT_FOUND)
+    data.resize(header.texSize);
+
+    if(header.texSize == header.texWidth * header.texHeight * header.bpp)
```
Do an early return, i.e. instead of
```
if(x)
{
    A;
}
else
{
    B;
    return;
}
```
do
```
if(!x)
{
    B;
    return;
}
A;
````

Quaker762 replied:
Done
